app/workers/generate_test_applications.rb

```diff
@@ -29,11 +31,17 @@ def perform
     create states: [:awaiting_provider_decision], apply_again: true
   end
 
-  def create(states:, apply_again: false)
+  def create(states:, apply_again: false, course_full: false)
     @test_applications.create_application(
       courses_to_apply_to: @courses_to_apply_to,
       states: states,
       apply_again: apply_again,
     )
+
+    if course_full
+      application = ApplicationForm.last
+      choice = application.application_choices.first
+      choice.course.course_options.update_all(vacancy_status: 'no_vacancies')
```

Comment:
No your original assessment was correct - enough runs of this task could result in all courses being full, which is a confusing set of side effects. 

There are a few potential alternatives I've considered:

- Use a unique course per ApplicationChoice being made. This puts a hard limit on the number of times you can run the task though, and I can see us potentially running into that limit quite quickly.
- Put a support button on an application form that allows the dev-support user to 'fill' the course being applied to. This could also feasibly result in all courses being filled if used enough times, but it's a lot slower and unlikely to happen by accident between database resets.
- Always use the same course as the 'full' course with each run of the task.

WDYT of the latter two? The third seems like the lightest implementation with minimal side effects.

Reply:
@malcolmbaig let's always use the same course, option 3 👍

Reply:
@tvararu added in latest commit